src/shared/watchdog.c

```diff
@@ -48,25 +48,42 @@ static int disable_watchdog(void) {
 }
 
 static int settimeout_watchdog(void) {
+        bool op_not_supported = false;
         usec_t t;
-        int sec;
+        int sec, r;
 
         assert(watchdog_fd > 0);
         assert(timestamp_is_set(watchdog_timeout));
 
         t = DIV_ROUND_UP(watchdog_timeout, USEC_PER_SEC);
         sec = MIN(t, (usec_t) INT_MAX); /* Saturate */
 
-        if (ioctl(watchdog_fd, WDIOC_SETTIMEOUT, &sec) < 0)
-                return log_warning_errno(errno, "Failed to set timeout to %is: %m", sec);
+        r = ioctl(watchdog_fd, WDIOC_SETTIMEOUT, &sec);
+        if (r < 0) {
+                /* Setting the timeout might not be supported by the HW. In
+                 * that case we'll reuse the effective timeout. */
+                if (!ERRNO_IS_NOT_SUPPORTED(errno))
+                        return log_warning_errno(errno, "Failed to set timeout to %is: %m", sec);
+
+                if (ioctl(watchdog_fd, WDIOC_GETTIMEOUT, &sec) < 0)
+                        return log_warning_errno(errno, "Failed to request watchdog HW timeout: %m");
+
+                op_not_supported = true;
+        }
+
+        assert(sec > 0);/*  buggy driver ? */
 
-        /* Just in case the driver is buggy */
-        assert(sec > 0);
+        if (watchdog_timeout != sec * USEC_PER_SEC) {
+                /* watchdog_timeout stores the timeout used by the HW */
+                watchdog_timeout = sec * USEC_PER_SEC;
 
-        /* watchdog_timeout stores the timeout used by the HW */
-        watchdog_timeout = sec * USEC_PER_SEC;
+                if (op_not_supported) {
+                        log_warning("Modifying the watchdog timeout is not supported by the HW");
+                        log_warning("Reusing the hardware watchdog timeout: %is", sec);
```

Comment:
i don't think this should be at warning level, it's not really something to be concerned about. at most LOG_NOTICE, but I think LOG_INFO should be totally sufficient I think.

Also, please always log everything you want to log in one log call, don't split this up, since this should show up as a single log event, not multiple. It's totally fine to embedd a newline in a log call, if you want.

use FORMAT_TIMESPAN() for outputting the time value

Reply:
Fixed